integration-tests/bats/sql-patch.bats

```diff
@@ -0,0 +1,391 @@
+    dolt sql -q "ALTER TABLE parent DROP PRIMARY KEY;"
+    dolt sql -q "ALTER TABLE parent ADD PRIMARY KEY(id, id_ext);"
+
+    dolt sql -q "CALL DOLT_PATCH()"
+    run dolt sql -q "CALL DOLT_PATCH()"
+    [ "$status" -eq 0 ]
+    [ "${#lines[@]}" -eq 20 ]
+
+    match_diff_and_patch_results
+}
+
+@test "sql-patch: reconciles CHECK CONSTRAINTS" {
+    dolt sql <<SQL
+create table foo (
+    pk int,
+    c1 int,
+    CHECK (c1 > 3),
+    PRIMARY KEY (pk)
+);
+SQL
+
+    dolt sql -q "CALL DOLT_PATCH()"
+    run dolt sql -q "CALL DOLT_PATCH()"
+    [ "$status" -eq 0 ]
+    [ "${#lines[@]}" -eq 10 ]
+
+    match_diff_and_patch_results
+}
+
+@test "sql-patch: any error causing no data diff is shown as warnings." {
+        dolt sql <<SQL
+CREATE TABLE parent (
+    id int PRIMARY KEY,
+    id_ext int,
+    v1 int,
+    v2 text COMMENT 'tag:1',
+    INDEX v1 (v1)
+);
+CREATE TABLE child (
+    id int primary key,
+    v1 int
+);
+SQL
+    dolt commit -Am "add tables"
+    dolt sql -q "ALTER TABLE child ADD CONSTRAINT fk_named FOREIGN KEY (v1) REFERENCES parent(v1);"
+    dolt sql -q "insert into parent values (0, 1, 2, NULL);"
+    dolt sql -q "ALTER TABLE parent DROP PRIMARY KEY"
+    dolt sql -q "ALTER TABLE parent ADD PRIMARY KEY(id, id_ext);"
+
+    run dolt diff -r sql child
+    [ "$status" -eq 0 ]
+    [[ "$output" =~ "Incompatible schema change, skipping data diff for table 'child'" ]] || false
+    diff_output_0=${lines[0]}
+    diff_output_1=${lines[1]}
+
+    run dolt sql -q "CALL DOLT_PATCH('child'); SHOW WARNINGS;"
+    [ "$status" -eq 0 ]
+    [[ "${lines[1]}" =~ "statement" ]] || false
+    [[ "${lines[3]}" =~ "$diff_output_0" ]] || false
+    [[ "${lines[4]}" =~ "$diff_output_1" ]] || false
+    [[ "$output" =~ "Incompatible schema change, skipping data diff for table 'child'" ]] || false
+
+    run dolt diff -r sql parent
+    [ "$status" -eq 0 ]
+    [[ "$output" =~ "Primary key sets differ between revisions for table 'parent', skipping data diff" ]] || false
+    diff_output_0=${lines[0]}
+    diff_output_1=${lines[1]}
+
+    run dolt sql -q "CALL DOLT_PATCH('parent'); SHOW WARNINGS;"
+    [ "$status" -eq 0 ]
+    [[ "${lines[1]}" =~ "statement" ]] || false
+    [[ "${lines[3]}" =~ "$diff_output_0" ]] || false
+    [[ "${lines[4]}" =~ "$diff_output_1" ]] || false
+    [[ "$output" =~ "Primary key sets differ between revisions for table 'parent', skipping data diff" ]] || false
+}
+
+# either no arguments or give two commit revisions
+match_diff_and_patch_results() {
+    if [ -z "$1" ] && [ -z "$2" ]; then
+        run dolt diff -r sql
+    else
+        run dolt diff -r sql $1 $2
+    fi
+    [ "$status" -eq 0 ]
+    diff_array=( "${lines[@]}" )
+
+    if [ -z "$1" ] && [ -z "$2" ]; then
+        run dolt sql -q "CALL DOLT_PATCH()"
+    else
+        run dolt sql -q "CALL DOLT_PATCH('$1','$2')"
+    fi
+    [ "$status" -eq 0 ]
+    [[ "${lines[1]}" =~ "statement" ]] || false
+    patch_array=( "${lines[@]:3}" )
```

Comment:
Would be nice to do a quick assertion that the diff_array doesn't contain any extra elements (otherwise we could compare all the ones in patch_array and miss that diff_array had more elements). 

ChatGPT told me this should work 😉 (you'll just need to account for the two extra elements in patch_array)
`[ ${#array1[@]} -eq ${#array2[@]} ];`

Reply:
I think this might not work for cases where `dolt patch` prints out error message for a table that has schema diff but cannot get the data diff error message. I made it so that this error message comes out as warning for `DOLT_PATCH()`. I added length check for each test for `CALL DOLT_PATCH()` before matching the array results. This would be helpful to check expected number of rows.